### ISSUE TITLE
Compatibility wrapper makes OpenSSL.tsafe unimportable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5969]
|<img alt="mgood's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mgood reported|
|-|-|
|Trac ID|trac#5969|
|Type|defect|
|Created|2012-09-12 17:59:55Z|
|Branch|https://github.com/twisted/twisted/tree/delete-tsafe-5969|

The twisted compatibility wrapper for `OpenSSL.tsafe` prevents importing of the module it's meant to make Twisted-compatible.

Without importing Twisted, this works:
```
$ python -c 'from OpenSSL import tsafe'
```

After importing Twisted, this fails:
```
$ python -c 'import twisted; from OpenSSL import tsafe'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: cannot import name tsafe
```

The code in twisted.python.compat injects `sys.modules['OpenSSL.tsafe']`, which prevents actually importing `OpenSSL.tsafe`.

By adding the key to sys.modules, Python thinks that the module has already been imported, and doesn't set the attribute "tsafe" on the real "OpenSSL" module object like it would when you import it normally.

If you import it this way, the import itself seems to succeed, but the attribute is not defined:

```
Python 2.7.1 (r271:86832, Jun 25 2011, 05:09:01) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2335.15.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import twisted
>>> import OpenSSL.tsafe
>>> OpenSSL.tsafe
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'tsafe'
```

To work, the monkey patch would need applied to the OpenSSL object as well, such as:

```
import OpenSSL
OpenSSL.tsafe = sys.modules['OpenSSL.tsafe'] = tsafe
```

This would probably also need a fallback when OpenSSL is not installed.

Tested in virtualenv with:

```
Python 2.6.7 & 2.7.1 (Mac)
Twisted 12.2.0
pyOpenSSL 0.13
wsgiref 0.1.2
zope.interface 4.0.1
```

<details><summary>Searchable metadata</summary>

```
trac-id__5969 5969
type__defect defect
reporter__mgood mgood
priority__normal normal
milestone__ 
branch__branches_delete_tsafe_5969 branches/delete-tsafe-5969
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347472795000000 1347472795000000
changetime__1347480700000000 1347480700000000
version__None None
owner__exarkun exarkun
cc__mithrandi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35609]) Branching to 'delete-tsafe-5969'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

The original motivation for this code is no longer valid, if it ever was.  Recent versions of CPython provide no compelling reason to try to delay the import of `threading` or `thread`.  There was some belief that at some point in history, importing these modules made the runtime slower, even for single-threaded code.  Whether or not that was ever true, it isn't true anymore, and since name resolution is thread-based by default, even if it were true this code wouldn't help.

I straight up deleted it in the linked branch.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/delete-tsafe-5969).

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi set owner to @exarkun|
|-|-|

I see the build on lucid32-py2.6-select failed, but I don't understand the test failure at all, so it may be unrelated. Assuming that's not an issue, then +1 please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [35615]) Merge delete-tsafe-5969

Author: exarkun
Reviewer: mithrandi
Fixes: [#5969](https://github.com/twisted/twisted/issues/5969)

Delete the pyOpenSSL monkey-patching code in `twisted.python.compat` which fecklessly attempted
to delay the import of the `threading` and `thread` modules.

**Comment by twisted-trac** [comment]
|<img alt="mgood's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|mgood commented|
|-|-|

Deleting it works too :) Thanks!
